datasets/newsgroup/newsgroup.py

```diff
@@ -0,0 +1,171 @@
+
+# Lint as: python3
+"""20Newsgroup  dataset"""
+
+from __future__ import absolute_import, division, print_function
+
+import csv
+import os
+
+import nlp
+
+
+_CITATION = """
+    @inproceedings{Lang95
+    ,author = "Ken Lang"
+    ,title = "Newsweeder: Learning to filter netnews"
+    ,year = 1995
+    ,booktitle = "Proceedings of the Twelfth International Conference on Machine Learning"
+    ,pages = "331-339"
+    }
+ """
+
+_DESCRIPTION = """
+The 20 Newsgroups data set is a collection of approximately 20,000 newsgroup documents, partitioned (nearly) evenly across 
+20 different newsgroups. To the best of my knowledge, it was originally collected by Ken Lang, probably for his Newsweeder: 
+Learning to filter netnews paper, though he does not explicitly mention this collection. The 20 newsgroups collection has become 
+a popular data set for experiments in text applications of machine learning techniques, such as text classification and text clustering.
+"""
+
+_DOWNLOAD_URL = {
+    "bydate": "http://qwone.com/~jason/20Newsgroups/20news-bydate.tar.gz",
+    "19997": "http://qwone.com/~jason/20Newsgroups/20news-19997.tar.gz",
+    "18828": "http://qwone.com/~jason/20Newsgroups/20news-18828.tar.gz"
+}
+_NEWS_GROUPS = [
+    "comp.graphics",
+    "comp.os.ms-windows.misc",
+    "comp.sys.ibm.pc.hardware",
+    "comp.sys.mac.hardware",
+    "comp.windows.x", 
+    "rec.autos",
+    "rec.motorcycles",
+    "rec.sport.baseball",
+    "rec.sport.hockey",	
+    "sci.crypt",
+    "sci.electronics",
+    "sci.med",
+    "sci.space",
+    "misc.forsale",	
+    "talk.politics.misc",
+    "talk.politics.guns",
+    "talk.politics.mideast",	
+    "talk.religion.misc",
+    "alt.atheism",
+    "soc.religion.christian",
+]
+_VERSIONS = {
+    "19997": "1.0.0",
+    
+    "bydate": "2.0.0",
+    "18828": "3.0.0"
+}
+
+_DESC = {
+    "19997": "the original, unmodified version.",
+    "bydate": "sorted by date into training(60%) and test(40%) sets, does not include cross-posts (duplicates) and does not include newsgroup-identifying headers (Xref, Newsgroups, Path, Followup-To, Date)",
+    "18828": 'does not include cross-posts and includes only the "From" and "Subject" headers.'
+}
+_CONFIG_NAMES = []
+for version in _VERSIONS:
+    for group in _NEWS_GROUPS:
+        _CONFIG_NAMES.append(version+'_'+group)
+        
+_CONFIG_NAMES = sorted(_CONFIG_NAMES)
+class NewsgroupConfig(nlp.BuilderConfig):
+    """BuilderConfig for 20Newsgroup."""
+
+    def __init__(self, data_dir, **kwargs):
+        """Constructs a 20Newsgroup.
+
+        Args:
+        data_dirs: str
+        **kwargs: keyword arguments forwarded to super.
+        """
+
+        super(NewsgroupConfig, self).__init__(**kwargs)
+        self.data_dir = data_dir
+
+
+class Newsgroups(nlp.GeneratorBasedBuilder):
+    
+    BUILDER_CONFIGS = [
+        NewsgroupConfig(
+            name=name,
+            description = _DESC[name.split('_')[0]],
+            data_dir=name.split('_')[1],
```

Comment:
Maybe we can rename this variable ? We use `data_dir` as the directory for manually downloaded data.
Something like `sub_directory` for example ?